### PR TITLE
emoji panel when type :

### PR DESCRIPTION
**Pull request by MilkeeyCat**
image
This code adds such a panel where you can see emojis which match your text. You can use up and down arrow keys to select an emoji you need. To select the emoji press Tab. There's one bug and it happens when you try to write an emoji, for example :troll and press tab and then autocompele come in and modifies data. Any ideas how to fix it?

### CODE REVIEW

**Comment by MilkeeyCat** [comment]
upd. i removed auto complete for emotes and now everything works :)

**Comment by MilkeeyCat** [comment]
upd. now fr everything works

**Comment by ChillerDragon** [comment]
That is amazing! Thanks for working on that.

Since you are the first person implementing a plugin. How did it go? Anything weird about the api? Any suggestions?

**Comment by ChillerDragon** [comment]
Sadly so far it feels like a downgrade to me. Typing for example ``:j<tab>`` used to autocomplete ``:justatest:`` for me.
I think it makes sense that you removed the old auto completion because they would kinda conflict with each other (thats also how slack for example seems to do it)

But it would be nice to keep the old way of doing it when this plugin is inactive. Ideally the old way would be moved to its own plugin. So users with slow internet or computers or people that like it simple. Can use the non preview way of completion. And others can turn your fancy preview on. We probably need to introduce a way to properly conflict plugins. As in saying maybe with what type of plugins or what specific plugin names one plugin will conflict. So a user trying to turn both of them on gets an alert saying the plugins should not be combined.

**Comment by MilkeeyCat** [comment]
As an easy fix i can remove a limitation to show emojis only after 2 chars. So then you would get justatest as expected :)

**Comment by MilkeeyCat** [comment]
> That is amazing! Thanks for working on that.
> 
> Since you are the first person implementing a plugin. How did it go? Anything weird about the api? Any suggestions?

Generally everything was good. I felt the same as when I just started making websites and there wasnt those fancy shmancy frontend web frameworks. Pretty good.

File: src/plugins/core/typing_emoji_panel.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        } else if (e.key == 'Tab' && this.isVisible) {
+            e.preventDefault();
+            this.insertEmoji(this.currentEmoji);
+        }
+
+        this.update();
+    }
+
+    selectNextEmoji() {
+        let index = this.emojis.indexOf(this.currentEmoji);
```
Comment (ChillerDragon) [inline src/plugins/core/typing_emoji_panel.ts:118]: Could be ``const index = ..`` the in a few other places too. At some point we should setup linting to force the usage of const if the variable is not mutated. Lets try to already use const so the diff later is not too big.

**Comment by ChillerDragon** [comment]
> As an easy fix i can remove a limitation to show emojis only after 2 chars. So then you would get justatest as expected :)

Sounds good. Has that any disadvantage? Why did you introduce the minimum 2 char requirement in the first place?

**Pull request merged**